### PR TITLE
Complete Support for DCU and ROCm

### PR DESCRIPTION
**Pull request by lfengad**
- Fix ROCM and DCU compilation for newly rebased TF version
- Support ROCm Bazel compilation flow for tao bridge
- Fix atomic operation CodeGen issue for MI210
- Add custom library ROCm ad DCU compilation support

### CODE REVIEW

File: scripts/python/tao_build.py [hunk 1]
```diff
@@ -451,8 +472,9 @@ def bazel_build(target, flag=""):
             flag += ' --config=disc_mkldnn'
 
         bazel_build(TARGET_TAO_COMPILER_MAIN, flag=flag)
-        bazel_build(TARGET_DISC_OPT, flag=flag)
-        bazel_build(TARGET_DISC_REPLAY, flag=flag)
+        if not args.dcu and not args.rocm:
+            bazel_build(TARGET_DISC_OPT, flag=flag)
```
Comment (minminsun) [inline scripts/python/tao_build.py:476]: Why disc-opt build is skipped for dcu/rocm? The disc-opt binary is required for disc FileCheck unit tests.

File: tao_compiler/mlir/disc/transforms/codegen_utils.h [hunk 2]
```diff
@@ -31,7 +31,12 @@ limitations under the License.
 namespace mlir {
 namespace disc_ral {
 
+#if TENSORFLOW_USE_ROCM
 constexpr int kWarpSize = 32;
```
Comment (minminsun) [inline tao_compiler/mlir/disc/transforms/codegen_utils.h:35]: The warp size of ROCM is the same as CUDA here?

File: tao_compiler/mlir/disc/transforms/codegen_utils.h [hunk 3]
```diff
@@ -84,7 +89,11 @@ constexpr const char* kThreadPerBlockHint = "disc_thread_per_block_hint";
 constexpr const int kRowReductionScheduleTurningSize = 512;
 
 // default num of threads per block used when doing codegen
+#if TENSORFLOW_USE_ROCM
 constexpr const int kThreadsRowReduction = 256;
```
Comment (minminsun) [inline tao_compiler/mlir/disc/transforms/codegen_utils.h:93]: Did you intend to use a different number of threads for row reduction?

File: tao_compiler/mlir/disc/transforms/codegen_utils.h (hunk 2 above)
Comment (lfengad) [reply tao_compiler/mlir/disc/transforms/codegen_utils.h:35]: Warp size should be 64 in ROCm, but modifying to 64 here naively will cause some further incompatible issue in the DISC CodeGen process. We will revisit this number setting later. Leaving it as 32 now can make current CodeGen okay.

**Review by minminsun — APPROVED** [review]
LGTM

**Pull request merged**